nx/lib/nx/shared.ex

```diff
@@ -377,16 +377,19 @@ defmodule Nx.Shared do
   @doc """
   Builds the type of an element-wise binary operation.
   """
-  def binary_type(a, b) when is_number(a) and is_number(b), do: Nx.Type.infer(a + b)
-  def binary_type(a, b) when is_number(a), do: Nx.Type.merge_number(type(b), a)
-  def binary_type(a, b) when is_number(b), do: Nx.Type.merge_number(type(a), b)
+  def binary_type(a, b) when (is_number(a) or is_boolean(a)) and (is_number(b) or is_boolean(b)),
+    do: Nx.Type.infer(a + b)
+
+  def binary_type(a, b) when is_number(a) or is_boolean(a), do: Nx.Type.merge_number(type(b), a)
+  def binary_type(a, b) when is_number(b) or is_boolean(b), do: Nx.Type.merge_number(type(a), b)
```

Comment:
We should revert these changes too, as the code here is meant to work with tensors and true/false are not tensor equivalents.